src/jit/gentree.cpp

```diff
@@ -12019,32 +12020,130 @@ GenTreePtr Compiler::gtFoldExprSpecial(GenTreePtr tree)
+                        // is GT_COMMAs enclosing the GT_ASG we are
+                        // looking for.
+                        JITDUMP(" bailing; unexpected copy op %s\n", GenTree::OpName(copy->gtOper));
+                    }
+                    goto FAIL;
+                }
+
+                // If the copy is a struct copy, make sure we know how to isolate
+                // any source side effects.
+                GenTreePtr copySrc          = copy->gtOp.gtOp2;
+                bool       hasSrcSideEffect = false;
+                bool       isStructCopy     = false;
+
+                if (gtTreeHasSideEffects(copySrc, GTF_SIDE_EFFECT))
+                {
+                    hasSrcSideEffect = true;
+
+                    if (copySrc->gtType == TYP_STRUCT)
+                    {
+                        isStructCopy = true;
+
+                        if ((copySrc->gtOper != GT_OBJ) && (copySrc->gtOper != GT_IND) && (copySrc->gtOper != GT_FIELD))
+                        {
+                            // We don't know how to handle other cases, yet.
+                            JITDUMP(" bailing; unexpected copy source struct op with side effect %s\n",
+                                    GenTree::OpName(copySrc->gtOper));
+                            goto FAIL;
+                        }
+                    }
+                }
+
+                // Proceed with the optimization
+                //
+                // Change the assignment expression to a NOP.
+                JITDUMP("\nBashing NEWOBJ [%06u] to NOP\n", dspTreeID(asg));
                 asg->gtBashToNOP();
 
-                op = gtNewIconNode(oper == GT_NE);
+                // Change the copy expression so it preserves key
+                // source side effects.
+                JITDUMP("\nBashing COPY [%06u]", dspTreeID(copy));
+
+                if (!hasSrcSideEffect)
+                {
+                    // If there were no copy source side effects just bash
+                    // the copy to a NOP.
+                    copy->gtBashToNOP();
+                    JITDUMP(" to NOP\n");
+                }
+                else if (!isStructCopy)
+                {
+                    // For scalar types, go ahead and produce the
+                    // value as the copy is fairly cheap and likely
+                    // the optimizer can trim things down to just the
+                    // minimal side effect parts.
+                    copyStmt->gtStmt.gtStmtExpr = copySrc;
+                    JITDUMP(" to scalar read via [%06u]\n", dspTreeID(copySrc));
+                }
+                else
+                {
+                    // For struct types read the first byte of the
+                    // source struct; there's no need to read the
+                    // entire thing, and no place to put it.
+                    assert(copySrc->gtOper == GT_OBJ || copySrc->gtOper == GT_IND || copySrc->gtOper == GT_FIELD);
```

Comment:
Am going to change the struct case to read a 32 bit unsigned value, so we end up with slightly smaller encodings on x86/x64; should also mitigate any worries about byteable dests.
```asm
      458B11               mov      r10d, dword ptr [r9]
```
instead of
```asm
       4D0FBE11             movsx    r10, byte  ptr [r9]
```